### ISSUE TITLE
Feature request: widgets for document image understanding and document image VQA

### ISSUE THREAD

**Issue by NielsRogge** [huggingface/huggingface_hub#286]
I'm currently adding LayoutLMv2 and LayoutXLM to HuggingFace Transformers. These models, built by Microsoft, have impressive capabilities for understanding document images (scanned documents, such as PDFs). LayoutLM is an extension of BERT that incorporates layout information, besides just text.

It would be really cool to have inference widgets for the following tasks:
* document image understanding 
* document image visual question answering

## Document image understanding
Document image understanding (also called form understanding) means, understanding all pieces of information of a document image. Example datasets here are FUNSD, CORD, SROIE and Kleister-NDA.

The input is a document image:
<img width="607" alt="Schermafbeelding 2021-08-19 om 10 12 18" src="https://user-images.githubusercontent.com/48327001/130034903-13f88eaf-7a6e-4dca-9973-e1e83a41deba.png">

The output should be the same image, but with colored bounding boxes, indicating for example which text are questions (blue), which are answers (blue), which are headers (orange), etc.
<img width="606" alt="Schermafbeelding 2021-08-19 om 10 12 45" src="https://user-images.githubusercontent.com/48327001/130035118-39cce15b-a9a7-4e5a-9c5c-ce759b89d272.png">

LayoutLMv2 solves this as a NER problem, using `LayoutLMv2ForTokenClassification`. First, an OCR engine is run on the image to get a list of words + corresponding coordinates. These are then tokenized, and together with the image sent through the LayoutLMv2 model. The model then labels each token using its classification head.

## Document visual question answering
Document visual question answering means, given an image + question, generate (or extract) an answer. Example datasets here are DocVQA.

LayoutLMv2 solves this as a extractive question answering problem similar to SQuAD. I've defined a `LayoutLMv2ForQuestionAnswering`, which predicts the `start_positions` and `end_positions`.

However, I don't think we can leverage the existing 'token-classification' and 'question-answering' pipelines, as the inputs are quite different (document images instead of text).

**Comment by osanseviero** [comment]
Thanks for your ideas! I think these are great

cc @mishig25 that has been working on adding new widgets. The first step for having the widget is to actually have this working in the Inference API.

We usually try to have the widgets be as generic as possible, so let's think in ways that this can be reused by similar tasks so we don't end up with 200 tasks :) 

For image classification, isn't the existing `image-classification` widget good enough? ([example](https://huggingface.co/osanseviero/keras-dog-or-cat)).

@Narsil deeply understand the inference API using `transformers`, so he can probably give good insights in how to integrate this for `transformers`.

**Comment by NielsRogge** [comment]
> For image classification, isn't the existing image-classification widget good enough? (example).

We can perhaps leverage it, the only difference is that this model requires a processor instead of a feature extractor for preparing the inputs.

> The first step for having the widget is to actually have this working in the Inference API.

This means, first defining a pipeline for each task?

**Comment by osanseviero** [comment]
> This means, first defining a pipeline for each task (if the existing ones can't be used)?

Yes. To give you a quick summary, we have `api-inference` which uses transformers and `api-inference-community` which is used for 3rd party libraries + wip generic inference (which we could use for a proof of concept). 

Each widget maps to a single task and the input/output specifications are somewhat similar between the the two inferences. In the case of `api-inference`, each task maps to a single `transformers` pipeline, but maybe @Narsil can correct me here. So if we want to have a working widget that uses `transformers`, we'll need to define the pipeline. For non-transformers we have more flexibility and could get something up relatively quickly.

**Comment by Narsil** [comment]
`image-classification` already exists:

- `image-token-classification` can be added
- `image-question-answering` too.

We can setup something quick&dirty with `generic` to demo the widget, however, proper integration within `transformers` seems like a great idea, but will take more time (heavy refactor of pipelines + BC in transfomers means more time for review)

**Comment by osanseviero** [comment]
Agree with @Narsil here. We should build a quick&dirty impl with `generic`, but a proper integration will take some time.

I'm not sure I understand how `image-token-classification` (for document-image-understanding) differs in terms of inputs/outputs from object detection. Aren't they the same?

**Comment by Narsil** [comment]
Interesting idea to fuse it with `object-detection` (which I insist is exactly `image-segmentation` but let's not dwell on it).

For @NielsRogge , the output is `[{bounding_box: ... , "label": "XXXX", "instance": "Y", "parent"?: "Z"}, ....]. I feel like the label could be the NER tag, but we still need a field to print the written text otherwise it's not really useful. "text_content"? could be a new optional key.

Maybe the output is different enough for a separate pipeline here. It's on the edge IMO.

**Comment by NielsRogge** [comment]
Actually, the output is the same as object detection (a list of class labels and bounding boxes). Of course, postprocessing is quite different (for DETR, one can use `DetrFeatureExtractor.post_process()`, for LayoutLMv2, I'm currently converting them manually in a cell in a Jupyter notebook, but I could add a similar `post_process()` method to the processor).

We can then probably just use the existing `image-classification` and (WIP) `object-detection` pipelines. The only difference is that now a `Processor` will be required for preparing the inputs + postprocessing instead of `FeatureExtractor`.

For the other task, DocVQA, a new pipeline will be required: `image-question-answering`.

**Comment by Narsil** [comment]
AFAIK, `Processor` is only a superset of `FeatureExtractor` which encompasses all lower-level objects (`FeatureExtractor` + `Tokenizer`) is that true, or is there some subtlety I am missing ? Pipeline try to stay away for `Processor` if possible as it makes logic harder to follow (too much magic).

**Comment by NielsRogge** [comment]
> AFAIK, Processor is only a superset of FeatureExtractor which encompasses all lower-level objects (FeatureExtractor + Tokenizer) is that true, or is there some subtlety I am missing ?

That is only true for the processors defined for CLIP and Wav2Vec2. For those, the processor is just a wrapper around the FeatureExtractor and Tokenizer, and it can act as one of the two at any point in time. However, the processor of LayoutLMv2 is different in the sense that it applies the feature extractor and tokenizer in a sequence.

**Comment by Narsil** [comment]
Thanks for the clarification, will have to dig a bit into this.

**Comment by osanseviero** [comment]
related: this could be a nice library to integrate https://github.com/mindee/doctr

**Comment by NielsRogge** [comment]
There's another beautiful libarry for deep-learning based layout analysis: https://github.com/Layout-Parser/layout-parser 

They might be interested too.
